base/services/w32time/W32Time.c

```diff
@@ -0,0 +1,374 @@
+                 * Otherwise, if FALSE set the Local Time.
+                 * Call SetLocalTime twice to ensure correct results.
+                 */
+                if (SystemTime)
+                {
+                    Ret = SetSystemTime(lpSystemTime);
+                }
+                else
+                {
+                    Ret = SetLocalTime(lpSystemTime) &&
+                          SetLocalTime(lpSystemTime);
+                }
+
+                /*
+                 * For the sake of security, restore the previous status again
+                 */
+                if (previouspriv.PrivilegeCount > 0)
+                {
+                    AdjustTokenPrivileges(hToken,
+                                          FALSE,
+                                          &previouspriv,
+                                          0,
+                                          NULL,
+                                          0);
+                }
+            }
+        }
+        CloseHandle(hToken);
+    }
+
+    return Ret;
+}
+
+static DWORD
+GetIntervalSetting()
+{
+    HKEY hKey;
+    DWORD dwData;
+    DWORD dwSize = sizeof(DWORD);
+
+    dwData = 0;
+    if (RegOpenKeyExW(HKEY_LOCAL_MACHINE,
+                      L"SYSTEM\\CurrentControlSet\\Services\\W32Time\\TimeProviders\\NtpClient",
+                      0,
+                      KEY_QUERY_VALUE,
+                      &hKey) == ERROR_SUCCESS)
+    {
+        RegQueryValueExW(hKey,
+                             L"SpecialPollInterval",
+                             NULL,
+                             NULL,
+                             (LPBYTE)&dwData,
+                             &dwSize);
+        RegCloseKey(hKey);
+    }
+    return dwData;
+}
+
+
+int
+SetTime(VOID)
+{
+    ULONG ulTime;
+    LONG lRet;
+    HKEY hKey;
+    DWORD dwIndex = 0;
+    DWORD dwValSize;
+    DWORD dwNameSize;
+    WCHAR szValName[MAX_VALUE_NAME];
+    WCHAR szData[256];
+    WCHAR szDefault[256];
+
+
+    lRet = RegOpenKeyExW(HKEY_LOCAL_MACHINE,
+                        L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\DateTime\\Servers",
+                        0,
+                        KEY_QUERY_VALUE,
+                        &hKey);
+    if (lRet != ERROR_SUCCESS)
+    {
+        return 1;
+    }
+
+    while (TRUE)
+    {
+        dwValSize = MAX_VALUE_NAME * sizeof(WCHAR);
+        szValName[0] = L'\0';
+        lRet = RegEnumValueW(hKey,
+                             dwIndex,
+                             szValName,
+                             &dwValSize,
+                             NULL,
+                             NULL,
+                             (LPBYTE)szData,
+                             &dwNameSize);
+        if (lRet == ERROR_SUCCESS)
+        {
+            /* Get data from default reg value */
+            if (szValName[0] == UNICODE_NULL) // if this is the "(Default)" key
+            {
+                StringCbCopy(szDefault, wcslen(szData), szData);
```

Comment:
This misses the point of using StringCbCopy ;)
You need to tell it the size of the destination buffer (in bytes, otherwise use the Cch version) in order for it to be safe. I.e. the second parameter should be `sizeof(szDefault)` here.

Reply:
I am just learning to use these functions, so I appreciate the correction and hopefully I will remember it. I will fix this.